File: cli/src/syfthub_cli/completion.py
"""Shell completion functions for SyftHub CLI."""

from __future__ import annotations

import json
import time
from typing import Any, cast

from syfthub_cli.config import CONFIG_DIR, load_config

# Cache configuration
CACHE_FILE = CONFIG_DIR / ".completion_cache.json"
CACHE_TTL = 300  # 5 minutes


def _get_cached_endpoints() -> list[dict[str, Any]] | None:
    """Get endpoints from cache if valid."""
    if not CACHE_FILE.exists():
        return None
    try:
        data = json.loads(CACHE_FILE.read_text())
        if time.time() - data.get("timestamp", 0) < CACHE_TTL:
            return cast(list[dict[str, Any]], data.get("endpoints", []))
    except (json.JSONDecodeError, KeyError, OSError):
        pass
    return None


def _fetch_and_cache_endpoints() -> list[dict[str, Any]]:
    """Fetch endpoints from API and cache them."""
    try:
        from syfthub_sdk import AuthTokens, SyftHubClient

        config = load_config()
        client = SyftHubClient(base_url=config.hub_url, timeout=10)
        if config.access_token:
            client.set_tokens(
                AuthTokens(
                    access_token=config.access_token,
                    refresh_token=config.refresh_token or "",
                )
            )

        endpoints = []
        with client:
            for i, ep in enumerate(client.hub.browse()):
                endpoints.append(
                    {
                        "owner": ep.owner_username,
                        "name": ep.name,
                        "type": ep.type.value,
                        "description": ep.description or "",
                    }
                )
                if i >= 500:  # Limit for completion performance
                    break

        # Cache the results
        CONFIG_DIR.mkdir(parents=True, exist_ok=True)
        CACHE_FILE.write_text(
            json.dumps(
                {
                    "endpoints": endpoints,
                    "timestamp": time.time(),
                }
            )
        )
        return endpoints
    except Exception:
        # Silently fail - completion should never break the shell
        return []


def _get_endpoints() -> list[dict[str, Any]]:
    """Get endpoints, using cache if available."""
    cached = _get_cached_endpoints()
    if cached is not None:
        return cached
    return _fetch_and_cache_endpoints()


def complete_ls_target(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete target for ls command (username or user/endpoint).

    - If incomplete doesn't contain "/" -> complete usernames
    - If incomplete contains "/" -> complete endpoints for that user
    """
    try:
        endpoints = _get_endpoints()

        if "/" in incomplete:
            # User is typing an endpoint path like "alice/my-"
            user_prefix = incomplete.split("/")[0].lower()
            endpoint_prefix = (
                incomplete.split("/", 1)[1].lower() if "/" in incomplete else ""
            )

            completions = []
            for ep in endpoints:
                if ep["owner"].lower() == user_prefix:
                    full_path = f"{ep['owner']}/{ep['name']}"
                    if ep["name"].lower().startswith(endpoint_prefix):
                        # Return (value, help_text)
                        help_text = (
                            f"{ep['type']}: {ep['description'][:40]}"
                            if ep["description"]
                            else ep["type"]
                        )
                        completions.append((full_path, help_text))
            return completions
        else:
            # User is typing a username
            seen_users: dict[str, int] = {}
            for ep in endpoints:
                owner = ep["owner"]
                if owner.lower().startswith(incomplete.lower()):
                    seen_users[owner] = seen_users.get(owner, 0) + 1

            return [
                (f"{user}/", f"{count} endpoint(s)")
                for user, count in sorted(seen_users.items())
            ]
    except Exception:
        return []


def complete_model_endpoint(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete model endpoint paths (user/endpoint where type=model)."""
    try:
        endpoints = _get_endpoints()
        completions = []

        for ep in endpoints:
            if ep["type"] != "model":
                continue

            full_path = f"{ep['owner']}/{ep['name']}"
            if full_path.lower().startswith(incomplete.lower()):
                help_text = ep["description"][:50] if ep["description"] else "model"
                completions.append((full_path, help_text))

        return completions
    except Exception:
        return []


def complete_data_source(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete data source endpoint paths."""
    try:
        endpoints = _get_endpoints()
        completions = []

        for ep in endpoints:
            if ep["type"] not in ("data_source", "model_data_source"):
                continue

            full_path = f"{ep['owner']}/{ep['name']}"
            if full_path.lower().startswith(incomplete.lower()):
                help_text = ep["description"][:50] if ep["description"] else ep["type"]
                completions.append((full_path, help_text))

        return completions
    except Exception:
        return []


def complete_aggregator_alias(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete aggregator aliases from local config."""
    try:
        config = load_config()
        completions = []

        for alias, agg_config in config.aggregators.items():
            if alias.lower().startswith(incomplete.lower()):
                completions.append((alias, agg_config.url))

        return completions
    except Exception:
        return []


def complete_accounting_alias(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete accounting service aliases from local config."""
    try:
        config = load_config()
        completions = []

        for alias, acc_config in config.accounting_services.items():
            if alias.lower().startswith(incomplete.lower()):
                completions.append((alias, acc_config.url))

        return completions
    except Exception:
        return []


def complete_config_key(
    ctx: Any,  # noqa: ARG001
    incomplete: str,
) -> list[tuple[str, str]]:
    """Complete configuration keys."""
    # Static list of allowed keys with descriptions
    allowed_keys = {
        "default_aggregator": "Default aggregator alias",
        "default_accounting": "Default accounting service alias",
        "timeout": "Request timeout in seconds",
        "hub_url": "SyftHub API URL",
    }

    completions = []
    for key, description in allowed_keys.items():
        if key.lower().startswith(incomplete.lower()):
            completions.append((key, description))

    return completions

File: cli-go/internal/completion/completion.go
// Package completion provides dynamic shell completion for SyftHub CLI.
package completion

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/OpenMined/syfthub/cli-go/internal/config"
	"github.com/openmined/syfthub/sdk/golang/syfthub"
)

// Cache configuration.
const (
	CacheTTL = 5 * time.Minute
)

// CachedEndpoint represents a cached endpoint for completion.
type CachedEndpoint struct {
	Owner       string `json:"owner"`
	Name        string `json:"name"`
	Type        string `json:"type"`
	Description string `json:"description"`
}

// CompletionCache represents the cached completion data.
type CompletionCache struct {
	Endpoints []CachedEndpoint `json:"endpoints"`
	Timestamp int64            `json:"timestamp"`
}

// getCacheFile returns the path to the completion cache file.
func getCacheFile() string {
	return filepath.Join(os.Getenv("HOME"), ".syfthub", ".completion_cache.json")
}

// getCachedEndpoints returns cached endpoints if valid.
func getCachedEndpoints() []CachedEndpoint {
	cacheFile := getCacheFile()
	data, err := os.ReadFile(cacheFile)
	if err != nil {
		return nil
	}

	var cache CompletionCache
	if err := json.Unmarshal(data, &cache); err != nil {
		return nil
	}

	// Check if cache is still valid
	if time.Now().Unix()-cache.Timestamp > int64(CacheTTL.Seconds()) {
		return nil
	}

	return cache.Endpoints
}

// fetchAndCacheEndpoints fetches endpoints from API and caches them.
func fetchAndCacheEndpoints() []CachedEndpoint {
	cfg := config.Load()

	client, err := syfthub.NewClient(
		syfthub.WithBaseURL(cfg.HubURL),
		syfthub.WithTimeout(10*time.Second),
	)
	if err != nil {
		return nil
	}
	defer client.Close()

	if cfg.HasTokens() {
		refreshToken := ""
		if cfg.RefreshToken != nil {
			refreshToken = *cfg.RefreshToken
		}
		client.SetTokens(&syfthub.AuthTokens{
			AccessToken:  *cfg.AccessToken,
			RefreshToken: refreshToken,
		})
	}

	ctx := context.Background()
	var endpoints []CachedEndpoint

	iter := client.Hub.Browse(ctx, syfthub.WithPageSize(100))
	count := 0
	for iter.Next(ctx) {
		ep := iter.Value()
		endpoints = append(endpoints, CachedEndpoint{
			Owner:       ep.OwnerUsername,
			Name:        ep.Slug,
			Type:        string(ep.Type),
			Description: ep.Description,
		})
		count++
		if count >= 500 { // Limit for completion performance
			break
		}
	}

	// Cache the results
	cacheFile := getCacheFile()
	cacheDir := filepath.Dir(cacheFile)
	if err := os.MkdirAll(cacheDir, 0755); err == nil {
		cache := CompletionCache{
			Endpoints: endpoints,
			Timestamp: time.Now().Unix(),
		}
		if data, err := json.Marshal(cache); err == nil {
			_ = os.WriteFile(cacheFile, data, 0600)
		}
	}

	return endpoints
}

// getEndpoints returns endpoints, using cache if available.
func getEndpoints() []CachedEndpoint {
	if cached := getCachedEndpoints(); cached != nil {
		return cached
	}
	return fetchAndCacheEndpoints()
}

// CompleteLsTarget provides completion for the ls command target.
// - If incomplete doesn't contain "/" -> complete usernames
// - If incomplete contains "/" -> complete endpoints for that user
func CompleteLsTarget(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) > 0 {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	endpoints := getEndpoints()
	if endpoints == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	var completions []string

	if strings.Contains(toComplete, "/") {
		// User is typing an endpoint path like "alice/my-"
		parts := strings.SplitN(toComplete, "/", 2)
		userPrefix := strings.ToLower(parts[0])
		endpointPrefix := ""
		if len(parts) > 1 {
			endpointPrefix = strings.ToLower(parts[1])
		}

		for _, ep := range endpoints {
			if strings.ToLower(ep.Owner) == userPrefix {
				if strings.HasPrefix(strings.ToLower(ep.Name), endpointPrefix) {
					fullPath := ep.Owner + "/" + ep.Name
					desc := ep.Type
					if ep.Description != "" {
						if len(ep.Description) > 40 {
							desc = ep.Type + ": " + ep.Description[:40] + "..."
						} else {
							desc = ep.Type + ": " + ep.Description
						}
					}
					completions = append(completions, fullPath+"\t"+desc)
				}
			}
		}
	} else {
		// User is typing a username
		seenUsers := make(map[string]int)
		for _, ep := range endpoints {
			if strings.HasPrefix(strings.ToLower(ep.Owner), strings.ToLower(toComplete)) {
				seenUsers[ep.Owner]++
			}
		}

		for user, count := range seenUsers {
			desc := "1 endpoint"
			if count > 1 {
				desc = fmt.Sprintf("%d endpoints", count)
			}
			completions = append(completions, user+"/\t"+desc)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveNoSpace
}

// CompleteModelEndpoint provides completion for model endpoint paths.
func CompleteModelEndpoint(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) > 0 {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	endpoints := getEndpoints()
	if endpoints == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	var completions []string
	for _, ep := range endpoints {
		if ep.Type != "model" {
			continue
		}

		fullPath := ep.Owner + "/" + ep.Name
		if strings.HasPrefix(strings.ToLower(fullPath), strings.ToLower(toComplete)) {
			desc := "model"
			if ep.Description != "" {
				if len(ep.Description) > 50 {
					desc = ep.Description[:50] + "..."
				} else {
					desc = ep.Description
				}
			}
			completions = append(completions, fullPath+"\t"+desc)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp
}

// CompleteDataSource provides completion for data source endpoint paths.
func CompleteDataSource(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) > 0 {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	endpoints := getEndpoints()
	if endpoints == nil {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	var completions []string
	for _, ep := range endpoints {
		if ep.Type != "data_source" && ep.Type != "model_data_source" {
			continue
		}

		fullPath := ep.Owner + "/" + ep.Name
		if strings.HasPrefix(strings.ToLower(fullPath), strings.ToLower(toComplete)) {
			desc := ep.Type
			if ep.Description != "" {
				if len(ep.Description) > 50 {
					desc = ep.Description[:50] + "..."
				} else {
					desc = ep.Description
				}
			}
			completions = append(completions, fullPath+"\t"+desc)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp
}

// CompleteAggregatorAlias provides completion for aggregator aliases.
func CompleteAggregatorAlias(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	cfg := config.Load()
	var completions []string

	for alias, agg := range cfg.Aggregators {
		if strings.HasPrefix(strings.ToLower(alias), strings.ToLower(toComplete)) {
			completions = append(completions, alias+"\t"+agg.URL)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp
}

// CompleteAccountingAlias provides completion for accounting service aliases.
func CompleteAccountingAlias(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	cfg := config.Load()
	var completions []string

	for alias, acc := range cfg.AccountingServices {
		if strings.HasPrefix(strings.ToLower(alias), strings.ToLower(toComplete)) {
			completions = append(completions, alias+"\t"+acc.URL)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp
}

// CompleteConfigKey provides completion for configuration keys.
func CompleteConfigKey(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) > 0 {
		return nil, cobra.ShellCompDirectiveNoFileComp
	}

	keys := map[string]string{
		"default_aggregator": "Default aggregator alias",
		"default_accounting": "Default accounting service alias",
		"timeout":            "Request timeout in seconds",
		"hub_url":            "SyftHub API URL",
	}

	var completions []string
	for key, desc := range keys {
		if strings.HasPrefix(strings.ToLower(key), strings.ToLower(toComplete)) {
			completions = append(completions, key+"\t"+desc)
		}
	}

	return completions, cobra.ShellCompDirectiveNoFileComp
}
